doc/source/dev/roadmap-detailed.rst

```diff
@@ -157,8 +173,8 @@ includes
       algorithms by Dierckx, de Boor, possibly other). 
 
 Once we have a reasonably feature complete set, we can start taking a long look
-at the future of the venerable FITPACK Fortran library, which currently is the
-only way of constructing smoothing splines in SciPy.
+at the future of the FITPACK Fortran library, which currently is the only way
+of constructing smoothing splines in SciPy.
```

Comment:
@ev-br et al, I would really love to see splines support for a) n-d `x` and b) other array API compatible backends. Is that in the cards?

Reply:
a) d>1 batches of the independent variable? Not sure TBH. Will complicate the interface at likely no perf gain. Would love to see a usage example where a python loop at call site does not cut it. Could you show one? Will be easier to play around with ergonomics. 

b) is definitely planned, yes! Effort most welcome. low level code is compiled though, so the GPU support will have to be limited to delegating (cupyx.scipy.interpolate is a thing). A technical thing to solve is how to do delegation for classes, not just functions.

Reply:
> Will complicate the interface...

How so? Instead of requiring 1D `x` and broadcasting it (implicitly) to the shape of `y`, we would accept n-d `x` and broadcast it to the shape of `y`. Seems natural to me coming from `stats`. (e.g. any paired sample statistic used with `axis`.)

> Would love to see a usage example where a python loop at call site does not cut it

It's mostly that there may be many call sites and it's undesirable to have loops and calls to `xp.asarray` at all of them.

The application I have in mind is approximating probability distribution functions with splines like in `scipy.stats.sampling`/UNURAN, but built with SciPy primitives. The advantage to implementing (a simplified version of) these techniques ourselves is that it would be much faster to generate the interpolating splines with vectorized function calls. (The old `NumericalInverseHermite` was way faster to instantiate than the new one, and only marginally slower at generating extremely large samples.)

As for efficiency, suppose `searchsorted` were to support N-D input. With that, I can imagine implementing an efficient piecewise linear interpolator for N-D `x` and `y` using only array API calls. With a linear solver that supports batch input, could the same be done for higher order splines?

Reply:
The analogy with pairs sample statistics only takes us that far, same for searchsorted: for an interpolation problem there's an inherent asymmetry betwen `x` and `y`. 

However this starts getting too abstract, so how about we try a real example, to see how it goes and what the actual speed bumps are.
So, `stats.sampling` --- let's start from `NumericalInverseHermite`, is this a good starting point?

Reply:
> That's the intended application, but I don't think it's useful for exploring the interface of such a interpolator.

why is that?

The example is cute, but to me it's quite firmly on the "best done at the call site" side. You just do exactly the same thing with any interpolator and that's it. And to my eyes the version with an calling a 1D interpolator twice or in a loop is more readable anyway.

Maybe there's a more "production"-grade example.

Reply:
> why is that?

Your concern was about the interface, but `NumericalInverseHermite` does not have the interface that is anything like what we're talking about?

> The example is cute,

?
You wrote "Will complicate the interface at likely no perf gain. "
The example wasn't supposed to be cute... it was to demonstrate the proposed interface of the interpolator. 

> Maybe there's a more "production"-grade example.

How does that affect the interface of the interpolator? Wasn't the concern - that the interface of the interpolator would be too complicated?

---

Perhaps more context would help?

With the old infrastructure, we can create frozen distributions with array parameters:

```python3
import numpy as np
import matplotlib.pyplot as plt
from scipy import stats
X = stats.norm(loc=[[0], [1]], scale=[[1], [2]])
x = np.stack((np.linspace(-3, 3, 300), np.linspace(-4, 6, 300)))
plt.plot(x.T, X.pdf(x).T)
plt.show()
```
![image](https://github.com/user-attachments/assets/630b6e39-48ab-4261-93b1-cb1ddd7f58af)

With the new infrastructure, we can do the same thing.

```python3
Y = stats.Normal(mu=[0, 1], sigma=[1, 2])
Y.plot()
plt.show()
```
![image](https://github.com/user-attachments/assets/19671894-36b2-4827-aab2-edd5d03173a3)

Part of the purpose of `stats.sampling` is to represent distribution functions with splines so that it is faster to work with them, but it doesn't work if the frozen distribution has array shape parameters.
```
from scipy.stats import sampling
sampling.NumericalInverseHermite(X)
# TypeError: only length-1 arrays can be converted to Python scalars
```

If we wanted something like UNURAN that works when the distribution has array shape parameters, the interpolator needs to supporting different `x` coordinates for each set of parameters.

(I'll anticipate a valid criticism of the intended application - the `x`-grids might want to be different lengths. Agreed, but it is a separate issue from the interface of the interpolators.)

Reply:
Thanks for the context, this is exactly what I meant to ask!
Let's continue in https://github.com/scipy/scipy/issues/22118